cub/cub/device/device_segmented_reduce.cuh

```diff
@@ -679,80 +667,78 @@ struct DeviceSegmentedReduce
+   * @param[out] d_out
    *   Pointer to the output aggregate
    *
-   * @param[in] num_segments  
+   * @param[in] num_segments
    *   The number of segments that comprise the sorting data
    *
-   * @param[in] d_begin_offsets  
-   *   Random-access input iterator to the sequence of beginning offsets of 
-   *   length `num_segments`, such that `d_begin_offsets[i]` is the first 
-   *   element of the *i*<sup>th</sup> data segment in `d_keys_*` and 
+   * @param[in] d_begin_offsets
+   *   Random-access input iterator to the sequence of beginning offsets of
+   *   length `num_segments`, such that `d_begin_offsets[i]` is the first
+   *   element of the *i*<sup>th</sup> data segment in `d_keys_*` and
    *   `d_values_*`
    *
-   * @param[in] d_end_offsets  
-   *   Random-access input iterator to the sequence of ending offsets of length 
-   *   `num_segments`, such that `d_end_offsets[i] - 1` is the last element of 
-   *   the *i*<sup>th</sup> data segment in `d_keys_*` and `d_values_*`.  
-   *   If `d_end_offsets[i] - 1 <= d_begin_offsets[i]`, the 
+   * @param[in] d_end_offsets
+   *   Random-access input iterator to the sequence of ending offsets of length
+   *   `num_segments`, such that `d_end_offsets[i] - 1` is the last element of
+   *   the *i*<sup>th</sup> data segment in `d_keys_*` and `d_values_*`.
+   *   If `d_end_offsets[i] - 1 <= d_begin_offsets[i]`, the
    *   *i*<sup>th</sup> is considered empty.
    *
-   * @param[in] stream  
-   *   **[optional]** CUDA stream to launch kernels within.  
+   * @param[in] stream
+   *   **[optional]** CUDA stream to launch kernels within.
    *   Default is stream<sub>0</sub>.
    */
   template <typename InputIteratorT,
             typename OutputIteratorT,
             typename BeginOffsetIteratorT,
             typename EndOffsetIteratorT>
-  CUB_RUNTIME_FUNCTION static cudaError_t
-  ArgMin(void *d_temp_storage,
-         size_t &temp_storage_bytes,
-         InputIteratorT d_in,
-         OutputIteratorT d_out,
-         int num_segments,
-         BeginOffsetIteratorT d_begin_offsets,
-         EndOffsetIteratorT d_end_offsets,
-         cudaStream_t stream = 0)
+  CUB_RUNTIME_FUNCTION static cudaError_t ArgMin(void *d_temp_storage,
+                                                 size_t &temp_storage_bytes,
+                                                 InputIteratorT d_in,
+                                                 OutputIteratorT d_out,
+                                                 int num_segments,
+                                                 BeginOffsetIteratorT d_begin_offsets,
+                                                 EndOffsetIteratorT d_end_offsets,
+                                                 cudaStream_t stream = 0)
   {
-    // Signed integer type for global offsets
-    using OffsetT = int;
+    // Integer type for global offsets
+    using OffsetT = detail::common_iterator_value_t<BeginOffsetIteratorT, EndOffsetIteratorT>;
```

Comment:
critical: I believe this is a breaking change. Let's consider the following code:

```cpp
thrust::device_vector<int> in(num_items, 1);
thrust::device_vector<cub::KeyValuePair<int, int>> out(num_segs);
thrust::device_vector<unsigned int> offsets {
  0,
  seg_size,
  2 * seg_size
};

cub::DeviceSegmentedReduce::ArgMax(d_temp_storage,
                                   temp_storage_bytes,
                                   in.begin(),
                                   out.begin(),
                                   num_segs,
                                   offsets.begin(),
                                   offsets.begin() + 1);
```

Regardless of the offset iterator value types, this code used to work as long as the output value type was convertible to `cub::KeyValuePair<int, int>`. Now the compilation fails when offset iterator value types are not `int` because `cub::KeyValuePair<int, int>` can't be constructed from `cub::KeyValuePair<T, int>`. I suggest we keep `OffsetT = int` in `Arg*` algorithms for now.